prime-router/src/main/kotlin/fhirengine/engine/FHIRDestinationFilter.kt

```diff
@@ -0,0 +1,561 @@
+     *        Result will be negated if [reverseFilter] is true
+     */
+    internal fun evaluateFilterAndLogResult(
+        filters: ReportStreamFilter,
+        bundle: Bundle,
+        reportId: ReportId,
+        actionHistory: ActionHistory,
+        receiver: Receiver,
+        filterType: ReportStreamFilterType,
+        defaultResponse: Boolean,
+        reverseFilter: Boolean = false,
+        focusResource: Base = bundle,
+        useOr: Boolean = false,
+    ): Boolean {
+        val evaluationFunction = if (useOr) ::evaluateFilterConditionAsOr else ::evaluateFilterConditionAsAnd
+        val (passes, failingFilterName) = evaluationFunction(
+            filters,
+            bundle,
+            defaultResponse,
+            reverseFilter,
+            focusResource
+        )
+        if (!passes) {
+            logFilterResults(
+                failingFilterName ?: "unknown",
+                bundle,
+                reportId,
+                actionHistory,
+                receiver,
+                filterType,
+                focusResource
+            )
+        }
+        return passes
+    }
+
+    /**
+     * Takes a [bundle] and [filter] and optionally a [focusResource], evaluates if the bundle passes all
+     * the filter conditions, or when [reverseFilter] is true, evaluates if the bundle doesn't pass at least one of the
+     * filter conditions.
+     * @param filter Filter that will be evaluated
+     * @param bundle FHIR Bundle that will be evaluated
+     * @param defaultResponse result when there are no filter conditions in the list
+     * @param reverseFilter Optional flag used to reverse the result of the filter evaluation
+     * @param focusResource Starting point for the evaluation, can be [bundle] if checking from root
+     * @return Pair: Boolean indicating if the bundle passes the filter or not
+     *         and String to use when logging the filter result
+     */
+    internal fun evaluateFilterConditionAsAnd(
+        filter: ReportStreamFilter?,
+        bundle: Bundle,
+        defaultResponse: Boolean,
+        reverseFilter: Boolean = false,
+        focusResource: Base = bundle,
+    ): Pair<Boolean, String?> {
+        if (filter.isNullOrEmpty()) {
+            return Pair(defaultResponse, "defaultResponse")
+        }
+        val failingFilters = mutableListOf<String>()
+        val exceptionFilters = mutableListOf<String>()
+        filter.forEach { filterElement ->
+            try {
+                val filterElementResult = FhirPathUtils.evaluateCondition(
+                    CustomContext(bundle, focusResource, shorthandLookupTable, CustomFhirPathFunctions()),
+                    focusResource,
+                    bundle,
+                    bundle,
+                    filterElement
+                )
+                if (!filterElementResult) failingFilters += filterElement
+            } catch (e: SchemaException) {
```

Comment:
I'm not sure we should be catching exceptions here.  The idea is that the filters here should be simple and low risk, if something throws an exception we want to fail loudly so we can fix it and then de-queue the message